govcd/adminvdc.go

```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
 
 	return updateAdminVdcStorageProfile, err
 }
+
+// AddStorageProfile adds a storage profile to a VDC
+func (vdc *AdminVdc) AddStorageProfile(storageProfile *types.VdcStorageProfileConfiguration, description string) (Task, error) {
+	if vdc.client.VCDHREF.String() == "" {
+		return Task{}, fmt.Errorf("cannot add VDC storage profile, VCD HREF is unset")
+	}
+
+	href := vdc.AdminVdc.HREF + "/vdcStorageProfiles"
+
+	var updateStorageProfile = types.UpdateVdcStorageProfiles{
+		Xmlns:                types.XMLNamespaceVCloud,
+		Name:                 storageProfile.ProviderVdcStorageProfile.Name,
+		Description:          description,
+		AddStorageProfile:    storageProfile,
+		RemoveStorageProfile: nil,
+	}
+
+	task, err := vdc.client.ExecuteTaskRequest(href, http.MethodPost,
+		types.MimeUpdateVdcStorageProfiles, "error adding VDC storage profile: %s", &updateStorageProfile)
+	if err != nil {
+		return Task{}, fmt.Errorf("cannot add VDC storage profile, error: %s", err)
+	}
+
+	return task, nil
+}
+
+// AddStorageProfileWait adds a storage profile to a VDC and return a refreshed VDC
+func (vdc *AdminVdc) AddStorageProfileWait(storageProfile *types.VdcStorageProfileConfiguration, description string) error {
+	task, err := vdc.AddStorageProfile(storageProfile, description)
+	if err != nil {
+		return err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return err
+	}
+	return vdc.Refresh()
+}
+
+// RemoveStorageProfile remove a storage profile from a VDC
+func (vdc *AdminVdc) RemoveStorageProfile(storageProfileName string) (Task, error) {
+	if vdc.client.VCDHREF.String() == "" {
+		return Task{}, fmt.Errorf("cannot remove VDC storage profile: VCD HREF is unset")
+	}
+
+	var storageProfile *types.Reference
+	for _, sp := range vdc.AdminVdc.VdcStorageProfiles.VdcStorageProfile {
+		if sp.Name == storageProfileName {
+			storageProfile = sp
+		}
+	}
+	if storageProfile == nil {
+		return Task{}, fmt.Errorf("cannot remove VDC storage profile: storage profile '%s' not found in VDC", storageProfileName)
+	}
+
+	vdcStorageProfileDetails, err := vdc.client.GetStorageProfileByHref(storageProfile.HREF)
+	if err != nil {
+		return Task{}, fmt.Errorf("cannot retrieve VDC storage profile '%s' details: %s", storageProfileName, err)
+	}
+	if vdcStorageProfileDetails.Enabled {
+		_, err = vdc.UpdateStorageProfile(extractUuid(storageProfile.HREF), &types.AdminVdcStorageProfile{
+			Name:    storageProfile.Name,
+			Units:   "MB",
```

Comment:
why not using vdcStorageProfileDetails.units ?

Reply:
changed